components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/PasswordGrantHandler.java

```diff
@@ -105,18 +105,28 @@ private void setPropertiesForTokenGeneration(OAuthTokenReqMessageContext tokReqM
     private boolean validateUserTenant(OAuth2AccessTokenReqDTO tokenReq, ServiceProvider serviceProvider)
             throws IdentityOAuth2Exception {
 
-        String userTenantDomain = MultitenantUtils.getTenantDomain(tokenReq.getResourceOwnerUsername());
+        String userTenantDomain = null;
+        String usernameFromTokenRequest = tokenReq.getResourceOwnerUsername();
+
         if (IdentityTenantUtil.isTenantQualifiedUrlsEnabled()) {
             if (!serviceProvider.isSaasApp() ||
                     (serviceProvider.isSaasApp() && !IdentityTenantUtil.isLegacySaaSAuthenticationEnabled())) {
+                String userNameWithTenant = usernameFromTokenRequest;
                 userTenantDomain = IdentityTenantUtil.getTenantDomainFromContext();
-                String tenantAwareUsername =
-                        MultitenantUtils.getTenantAwareUsername(tokenReq.getResourceOwnerUsername());
-                String userNameWithTenant =
-                        tenantAwareUsername + UserCoreConstants.TENANT_DOMAIN_COMBINER + userTenantDomain;
+
+                if (!usernameFromTokenRequest.endsWith(userTenantDomain)) {
```

Comment:
Still there can be an issue in IdentityTenantUtil.isTenantQualifiedUrlsEnabled() && serviceProvider.isSaasApp() & IdentityTenantUtil.isLegacySaaSAuthenticationEnabled()) mode. Please check that mode.